File: app/src/main/java/com/fairshare/domain/model/sync/Operation.kt
package com.fairshare.domain.model.sync

import kotlinx.serialization.Serializable

/**
 * A single CRDT operation. Immutable: once emitted, the envelope is appended to
 * the local op log and never mutated.
 *
 * Convergence is achieved via a Lamport pair `(lamport, deviceId)` (DESIGN.md
 * §4): the materializer keeps, per entity, the surviving op with the highest
 * pair lexicographically.
 *
 * @property opId Globally unique UUID v4 generated at emit time. Doubles as the
 *   idempotency key when transports redeliver an op.
 * @property eventId Scopes the op to one Event; cross-event ops do not exist.
 * @property deviceId Emitter device, also the LWW tiebreaker.
 * @property lamport Per-device monotonic counter; on receive `local = max(local,
 *   received) + 1` (DESIGN.md §4.1).
 * @property wallClockMs Best-effort wall clock for debugging only. Not used by
 *   the conflict resolution algorithm.
 * @property payload The mutation itself.
 */
@Serializable
data class Operation(
    val opId: String,
    val eventId: String,
    val deviceId: String,
    val lamport: Long,
    val wallClockMs: Long,
    val payload: OpPayload,
) {
    /**
     * Lexicographic comparator over `(lamport, deviceId)`. Higher values win.
     * Used by the materializer to pick the surviving op per entity.
     */
    companion object {
        val LwwOrder: Comparator<Operation> = Comparator { a, b ->
            val byLamport = a.lamport.compareTo(b.lamport)
            if (byLamport != 0) byLamport else a.deviceId.compareTo(b.deviceId)
        }
    }
}

File: webapp/src/core/sync/operation.ts
import type { OpPayload } from "./opPayload";

/**
 * A single CRDT operation. Immutable: once emitted, persisted to the
 * local op log and never mutated. Convergence comes from the
 * `(lamport, deviceId)` Lamport pair — see `materializer.ts`.
 *
 * Mirrors `com.fairshare.domain.model.sync.Operation`. The on-wire JSON
 * field order is exactly: opId, eventId, deviceId, lamport, wallClockMs,
 * payload — declaration order from the Kotlin data class.
 */
export interface Operation {
  opId: string; // UUID v4, also the idempotency key
  eventId: string;
  deviceId: string;
  lamport: number;
  wallClockMs: number;
  payload: OpPayload;
}

/**
 * Lexicographic compare over `(lamport, deviceId)`. Returns a negative
 * number when `a < b`, positive when `a > b`, zero on equality. The
 * materializer picks the *maximum* per entity, so this is wired into
 * `Array.prototype.sort` callbacks accordingly.
 */
export function compareLww(a: Operation, b: Operation): number {
  if (a.lamport !== b.lamport) return a.lamport - b.lamport;
  if (a.deviceId < b.deviceId) return -1;
  if (a.deviceId > b.deviceId) return 1;
  return 0;
}
